### PR TITLE
internal/gocdk: consolidate materialize mechanisms & introduce sharing

### LINKED ISSUE #2327: internal/gocdk: consolidate ways we materialize static assets
`init`, `biome add`, `demo add`, and `provision add` all take files from the static content and instantiate them, but they all do it in different ways. Consolidate.

### PR DESCRIPTION
**Pull request by vangent**
*  All of `init`, `biome add`, `demo add`, and `provision add` now use functions from the `static` package to materialize their changes. `static.Open` is no longer exported.

* Instead of the `.json` config file, I ended up using the same struct, but filling it in with code. This is more flexible and easier to read as the directives get more complicated. For example, `static` has helpers like "add a Terraform provider", so I was able to delete most of the content from the `provision` types.

Fixes #2327.

### CODE REVIEW

**Comment by zombiezen** [comment]
Eep, sorry I sent out the review too soon while admiring the VSCode Pull Request extension.

File: internal/cmd/gocdk/internal/static/static.go [hunk 1]
```diff
@@ -14,12 +14,315 @@
...
+			continue
+		}
+		retval = append(retval, info.Name())
+	}
+	return retval, nil
+}
+
+// Action represents an action relative to static content that will be performed
+// in Do.
+type Action struct {
```
Comment (zombiezen) [inline internal/cmd/gocdk/internal/static/static.go]: Do we expect for someone to create an Action outside of using the functions you provide? It seems like the functions are a more approachable way of creating these object.

File: internal/cmd/gocdk/internal/static/static.go [hunk 2]
```diff
@@ -14,12 +14,315 @@
...
+	}
+	return retval, nil
+}
+
+// Options holds options for Do.
+type Options struct {
+	// Force allows Do to clobber existing files when copying.
+	Force bool
+	// Logger is used to log information about changes.
+	Logger *log.Logger
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/cmd/gocdk/internal/static/static.go]: What interface? If we want to get the "gocdk: " prefix for stderr, we need to use the stderr logger. Could make this an `io.Writer` and use `pctx.stderr`, but then you'd lose that prefix.

File: internal/cmd/gocdk/internal/static/static.go [hunk 3]
```diff
@@ -14,12 +14,315 @@
...
+		}
+		// Move past the marker itself, and to the next line.
+		idx += len(markerBytes)
+		for existing[idx] == '\n' || existing[idx] == '\r' {
+			idx++
+		}
+	}
+	newContent := append(existing[:idx], append(content, existing[idx:]...)...)
+
+	// TODO(rvangent): Copy os.FileMode from the current file?
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/cmd/gocdk/internal/static/static.go]: OK, removing TODO.

File: internal/cmd/gocdk/internal/static/static.go [hunk 4]
```diff
@@ -14,12 +14,315 @@
...
+// destRoot.
+//
+// By default, it copies files. If a file already exists at the destination,
+// Do will return an error unless force is true.
+//
+// Additional behavior can be configured by adding a JSON file to the static
+// data, at srcRoot + ".json". The JSON file should consist of a sequence of
+// JSON objects that will be deserialized into fileOptions structs; see
+// the struct docstring for more info.
+func Do(destRoot string, actions []*Action, opts *Options) error {
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/cmd/gocdk/internal/static/static.go]: Cool, done.

File: internal/cmd/gocdk/internal/static/static.go [hunk 5]
```diff
@@ -14,12 +14,315 @@
...
+	}
+}
+
+// CopyDir returns a slice of Action that recursively copy all of the files
+// under srcRoot in the static data to the same paths relative to the destRoot
+// passed to Do.
+func CopyDir(srcRoot string) ([]*Action, error) {
+	files, err := ListFiles(srcRoot)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/cmd/gocdk/internal/static/static.go]: Maybe. If so, the `ListFiles` should panic. Added a TODO to consider panicing in cases where the static data is wrong.

File: internal/cmd/gocdk/internal/static/static.go (hunk 1 above)
Comment (vangent) [reply internal/cmd/gocdk/internal/static/static.go]: Maybe? Not sure yet.

File: internal/cmd/gocdk/internal/static/static.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zombiezen) [reply internal/cmd/gocdk/internal/static/static.go]:
```go
type Printer interface {
  Printf(format string, args ...interface{})
}
```

File: internal/cmd/gocdk/internal/static/static.go (hunk 1 above)
Comment (zombiezen) [reply internal/cmd/gocdk/internal/static/static.go]: I have a package that I wrote like this for gg: https://github.com/zombiezen/gg/blob/master/internal/filesystem/filesystem.go

I found it easier to keep the fields unexported and just use exported constructor functions to build the actions.

Comment (vangent) [reply internal/cmd/gocdk/internal/static/static.go]: That is definitely easier, and expected. I added a TODO to consider un-exporting the fields. At least `TemplateData` is used externally so would need to stay public.

File: internal/cmd/gocdk/internal/static/static.go [hunk 6]
```diff
@@ -14,12 +14,315 @@
...
+// addToFile adds content into the file at path.
+//
+// addToFile no-ops if skipMarker is found in the existing content of
+// the file at path. skipMarker defaults to content.
+//
+// If marker is the empty string, addToFile appends the content;
+// otherwise, it inserts it on the next line after the marker.
+//
+// It returns true if changes were made.
+func addToFile(path, marker, skipMarker string, content []byte) (bool, error) {
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/cmd/gocdk/internal/static/static.go]: Done.

File: internal/cmd/gocdk/internal/static/static.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (vangent) [reply internal/cmd/gocdk/internal/static/static.go]: OK.

**Comment by vangent** [comment]
PTAL.

**Comment by vangent** [comment]
@clausti I am going to submit so that I can get a bunch of stuff in before vacation. Please feel free to review and send comments on this PR or via issues (or however you like).

**Pull request merged**